### PR TITLE
Implement Add/Delete Timeslot functionality

### PR DESCRIPTION
**Pull request by appleJax**
Holy shit, I spent an embarrassing amount of time on this one. Finally got it working, except I could not figure out how to update the UI with the new timeslots without reloading the page with a very hacky `window.location.reload()`. We could probably fix it by making another ajax request and updating the props within the `then()` clause of the post request. It's a little trickier than normal because there are child components and modals involved. I can look into it more later.

Anyways, any suggestions on how to clean up the code would be appreciated. There's kind of a lot going on the way it's set up now.

Screenshots:
screen shot 2017-09-06 at 12 16 54 pm

screen shot 2017-09-06 at 11 55 03 am

screen shot 2017-09-06 at 11 56 55 am

screen shot 2017-09-06 at 11 57 06 am

screen shot 2017-09-06 at 12 00 25 pm

### CODE REVIEW

**Comment by gabrielbonner** [comment]
Maybe add an similar error display format for events in the past?  Currently it looks like this, which is fine, but you know, for consistencies sake:

screen shot 2017-09-06 at 7 40 44 pm

**Comment by appleJax** [comment]
@gabrielbonner fixed it. I added the `noValidate` attribute to the form, and I added a custom validation for timeslots in the past.

File: app/javascript/components/event.jsx [hunk 1]
```diff
@@ -1,51 +1,84 @@
...
-                  .map(user =>
-                    <th key={user.id} className='user'>{user.name}</th>
+import Row from './TimeslotRow'
+import AddTimeslot from './AddTimeslot'
+
+class Event extends Component {
+  constructor(props) {
+    super(props)
+
+    const {event, users, timeslots} = props
```
Comment (gabrielbonner) [inline app/javascript/components/event.jsx]: Nice, I keep meaning to use mass assignment but then I forget about it.

**Comment by gabrielbonner** [comment]
Nice freaking work.  I don't care how long it took you, would've taken me much longer!  :)

It is a fair amount of code but there's no refactoring jumping out at me that would pare it down.  You could use something like moment.js to reduce dateFormat.jsx, but I'm assuming you didn't do that for a reason.

That modal is nice, it would be interesting to see if we could pass in the text/button text/button functions as props so that we could reuse it, but that's definitely another ticket.

As for reloading the window, maybe we should looks at incorporating Redux?  That would solve re-rendering issues, it's a more realistic solution to what we encounter in the real world, and I don't think this app is too small to justify not using it.

And are there a few more tests that can be written with your changes to the timeslot.rb and user.rb models?

**Comment by gabrielbonner** [comment]
Maybe a clock spinner would be more apropos for the loading animation? https://codepen.io/andrewliebchen/pen/spBqc

**Comment by appleJax** [comment]
@gabrielbonner thanks! Yeah, I definitely think redux is worth looking into. I haven't looked at moment.js, I should probably do that. I thought about generalizing the modal to make it reusable, but I figured I would wait until we have more than two, so we can see which traits all the modals have in common.

Nice clock spinner! Yeah, we can play around with different options...

**Comment by jeremyjung** [comment]
Looking good so far.  I did some preliminary work to refresh the window on add/remove of timeslots.  It works, but there are a lot of opportunities to refactor here I think.

**Comment by jeremyjung** [comment]
I forgot I had an auto formatter on in my text editor so there are a bunch of white space changes and things like that that make the commit noisy.  Sorry about that.

**Comment by appleJax** [comment]
@jeremyjung fantastic! Great job, it's sooo much better without that nasty reload. Awesome idea abstracting out the api calls. It's starting to come together. I've just rearranged the code so many times that it felt messy to me, but looking at the whole after it's been cleaned up, it actually feels pretty nice. I'm happy with this functionality. Any additional suggestions are welcome though...

**Comment by appleJax** [comment]
Is this good to merge?

**Comment by gabrielbonner** [comment]
@appleJax I'll pull the latest changes and double check on my machine tonight, but I'm assuming that if it runs on yours it's good to go.

**Comment by gabrielbonner** [comment]
👍  on the merge

**Pull request merged**